### PR TITLE
feat: Add some useful generics

### PR DESCRIPTION
**Pull request by kentquirk**
## Which problem is this PR solving?

- Some features I plan to import from Refinery 3 will likely use the Fanout generic that was added there
- The gossip-based peer mechanism will need SetWithTTL

## Short description of the changes

- Create and test SetWithTTL
- Import code and tests for Fanout

### CODE REVIEW

File: generics/fanout_test.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+			mut.Lock()
+			cleanupTotal += i
+			mut.Unlock()
+		}
+		return worker, cleanup
+	}
+
+	result := Fanout(input, parallelism, workerFactory, nil)
+	assert.ElementsMatch(t, []int{2, 4, 6, 8, 10}, result)
+	assert.Equal(t, 6, cleanupTotal) // 0 + 1 + 2 + 3
```
Comment (VinozzZ) [inline generics/fanout_test.go:64]: Would it be easier to reason about to do `cleanupTotal += 1` and assert that `assert.Eqaul(t, parallelism, cleanupTotal)`?

File: generics/fanout_test.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+			mut.Lock()
+			cleanupTotal += i
+			mut.Unlock()
+		}
+		return worker, cleanup
+	}
+
+	result := FanoutToMap(input, parallelism, workerFactory, nil)
+	assert.EqualValues(t, expected, result)
+	assert.Equal(t, 6, cleanupTotal) // 0 + 1 + 2 + 3
```
Comment (VinozzZ) [inline generics/fanout_test.go:125]: Same as above

File: generics/setttl.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		TTL:   ttl,
+	}
+	s.Add(es...)
+	return s
+}
+
+// Add adds elements `es` to the SetWithTTL.
+func (s SetWithTTL[T]) Add(es ...T) {
+	for _, e := range es {
+		s.Items[e] = time.Now().Add(s.TTL)
```
Comment (VinozzZ) [inline generics/setttl.go:31]: If we still want to use `clockwork` package as a unified way to management time in Refinery, we will need to inject that into this generic implementation of SetWithTTL

File: generics/setttl.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	"cmp"
+	"sort"
+	"time"
+
+	"golang.org/x/exp/maps"
+)
+
+// SetWithTTL is a unique set of items with a TTL (time to live) for each item.
+// After the TTL expires, the item is automatically removed from the set.
+type SetWithTTL[T cmp.Ordered] struct {
```
Comment (VinozzZ) [inline generics/setttl.go:13]: Should we make this implementation concurrent safe? If not, maybe we can add a note about it in the comments

File: generics/setttl.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	maps.DeleteFunc(s.Items, func(k T, exp time.Time) bool {
+		return exp.Before(time.Now())
+	})
+}
+
+// Members returns the unique elements of the SetWithTTL in sorted order.
+// It also removes any items that have expired.
+func (s SetWithTTL[T]) Members() []T {
+	s.cleanup()
+	members := make([]T, 0, len(s.Items))
```
Comment (VinozzZ) [inline generics/setttl.go:61]: nit - we can use `maps.Keys` here

File: generics/setttl.go (hunk 3 above)
Comment (kentquirk) [reply generics/setttl.go:31]: We can't really do injection here; this is not a Start-able component. I also didn't want to make it dependent on passing a Clock in whenever you need one. 

What do you think of giving it a Clock member that defaults to a Real clock, and you have to manually override it in tests?

Comment (VinozzZ) [reply generics/setttl.go:31]: Yes, that's actually what I was thinking too

**Review by VinozzZ — APPROVED** [review]

**Pull request merged**